File: leetcode/0002-add-two-numbers.php
<?php

/**
 * 2. Add Two Numbers
 * https://leetcode.com/problems/add-two-numbers/
 */

/**
 * Definition for a singly-linked list.
 * class ListNode {
 *     public $val = 0;
 *     public $next = null;
 *     function __construct($val = 0, $next = null) {
 *         $this->val = $val;
 *         $this->next = $next;
 *     }
 * }
 */
class Solution
{
    function nodeToRevString($node)
    {
        return $node->val . ($node->next !== null ? $this->nodeToRevString($node->next) : '');
    }

    function compareRevStrings($s1, $s2)
    {
        $maxLength = max([strlen($s1), strlen($s2)]);

        $s3 = '';
        $val_next = 0;

        for ($i = 0; $i <= $maxLength; $i++) {
            $val = (int)($s1[$i] ?? 0) + (int)($s2[$i] ?? 0) + $val_next;
            $val_next = 0;
            if ($val >= 10) {
                $val -= 10;
                $val_next = 1;
            }
            $s3 .= $i < $maxLength || $val != 0 ? $val : '';
        }
        return $s3;
    }

    function revStringToNode($s)
    {
        return new ListNode($s[0], isset($s[1]) ? $this->revStringToNode(substr($s, 1)) : null);
    }

    /**
     * @param ListNode $l1
     * @param ListNode $l2
     * @return ListNode
     */
    function addTwoNumbers($l1, $l2)
    {
        $s1 = $this->nodeToRevString($l1);
        $s2 = $this->nodeToRevString($l2);
        $s3 = $this->compareRevStrings($s1, $s2);
        echo $s3;
        return $this->revStringToNode($s3);
    }
}

File: swift/0002-add-two-numbers.swift
/**
 * 2. Add Two Numbers
 * https://leetcode.com/problems/add-two-numbers/
 */


public class ListNode {
    public var val: Int
    public var next: ListNode?
    public init() { self.val = 0; self.next = nil; }
    public init(_ val: Int) { self.val = val; self.next = nil; }
    public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
}

class Solution {
    func addTwoNumbers(_ l1: ListNode?, _ l2: ListNode?) -> ListNode? {
        var list1 = l1
        var list2 = l2
        var oneInMemory = false
        let ans: ListNode? = .init(0)
        var current: ListNode? = ans
        
        while list1 != nil || list2 != nil || oneInMemory {
            let newValue = (list1?.val ?? 0) + (list2?.val ?? 0) + (oneInMemory ? 1 : 0)

            oneInMemory = newValue >= 10

            current!.next = .init(newValue % 10)
            current = current!.next

            list1 = list1?.next
            list2 = list2?.next
        }

        return ans!.next
    }
}